### PR TITLE
Implement groupUserProgress service

### PR DESCRIPTION
**Pull request by zenovich**
The second half of [groupTeamUserProgress](https://docs.google.com/document/d/15Xxmatrs_7my5ioriikL6YgOU4I0sAVcq5ySKinc0Nw/edit#heading=h.c7xpcaxby5ym)

### CODE REVIEW

File: app/api/groups/get_user_progress.go [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+		Where("groups.ID IN (?)", userGroupIDs).
+		Group("groups.ID, items.ID").
+		Order(gorm.Expr(
+			"FIELD(groups.ID"+strings.Repeat(", ?", len(userGroupIDs))+")",
+			userGroupIDs...)).
+		Order(gorm.Expr(
+			"FIELD(items.ID"+strings.Repeat(", ?", len(itemIDs))+")",
+			itemIDs...)).
+		Order("attempt_with_best_score.iMinusScore, attempt_with_best_score.sBestAnswerDate").
+		ScanIntoSliceOfMaps(&dbResult).Error())
```
Comment (smadbe) [inline app/api/groups/get_user_progress.go]: (cont. of what I wrote above)
Globally, I would have expected just take all the attempts from the group_id IN (user self id or all teams the user belongs to), and just do the max etc on this. Would prevent the 3 joins for each operations, would be much much simpler... but maybe that's less efficient (?).
Or I missed something 🙂

Comment (zenovich) [reply app/api/groups/get_user_progress.go]: That's exactly how I implemented the task in the beginning. The problem is that it doesn't work on large data sets (it didn't even finish in 15,000 seconds): ```SELECT ID FROM groups_attempts WHERE idGroup IN (...) AND idItem = ... ORDER BY idItem, iMinusScore, sBestAnswerDate LIMIT 1``` doesn't use indices at all. That's why I finally implemented the task this way. Yes it looks strange, but each of these three JOINs uses the index, so the whole query works really fast, it takes just about half a second to complete on the same large data set.

File: app/api/groups/get_user_progress.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+	service.MustNotBeError(srv.Store.GroupGroups().
+		Joins(`
+			JOIN groups_ancestors
+			ON groups_ancestors.idGroupChild = groups_groups.idGroupParent AND
+				groups_ancestors.idGroupAncestor != groups_ancestors.idGroupChild AND
+				groups_ancestors.idGroupAncestor = ?`, groupID).
+		Joins("JOIN groups ON groups.ID = groups_groups.idGroupParent AND groups.sType = 'Team'").
+		Where("groups_groups.idGroupChild IN (?)", userGroupIDs).
+		Where("groups_groups.sType IN ('direct', 'requestAccepted', 'invitationAccepted')").
+		Pluck("groups_groups.idGroupParent", &teamIDs).Error())
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/groups/get_user_progress.go]: > I don't get why you need to get the teams related to the input group as the request is about users and their results whatever the teams in which they did their best result.

It is done just to optimize the query. While joining team_links (user's teams) we consider only teams that are descendants of `group_id`, not all the user's parent groups.

[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/groups/get_user_progress.go]: fixed it to match the docs

**Comment by zenovich** [comment]
improved tests

File: app/api/groups/get_user_progress.go [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+)
+
+// swagger:operation GET /groups/{group_id}/user-progress users groups attempts items groupUserProgress
+// ---
+// summary: Display the current progress of users on a subset of items
+// description: For all children of items from the parent_item_id list,
+//              display the result of all user self-groups among the descendants of the given group
+//              (including those in teams).
+//
+//              For all users, only the attempt corresponding to the best score counts
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/groups/get_user_progress.go]: fixed

File: app/api/groups/get_user_progress.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+//              (across all his teams and his own attempts), disregarding whether or not
+//              the score was done in a team which is descendant of the input group.
+// parameters:
+// - name: group_id
+//   in: path
+//   type: integer
+//   required: true
+// - name: parent_item_ids
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/groups/get_user_progress.go]: fixed

File: app/api/groups/get_user_progress.feature [hunk 5]
```diff
@@ -0,0 +1,594 @@
...
+
+  Scenario: No visible items
+    Given I am the user with ID "1"
+    When I send a GET request to "/groups/1/user-progress?parent_item_ids=1010"
+    Then the response code should be 200
+    And the response body should be, in JSON:
+    """
+    [
+    ]
+    """
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/groups/get_user_progress.feature]: added tests

File: app/doc/responses.go [hunk 6]
```diff
@@ -162,3 +162,44 @@ type groupsGetTeamProgressResponse struct {
...
+	// in: body
+	Body []struct {
+		// The user’s self `group_id`
+		// required:true
+		GroupID int64 `json:"group_id,string"`
+		// required:true
+		ItemID int64 `json:"item_id,string"`
+		// The best score across all attempts, or 0
+		// required:true
+		Score float32 `json:"score"`
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/doc/responses.go]: fixed

**Pull request merged**